source/common/upstream/outlier_detection_impl.h

```diff
@@ -43,16 +81,43 @@ class OutlierDetectorHostSinkImpl : public OutlierDetectorHostSink {
  */
 class OutlierDetectorImpl : public OutlierDetector {
 public:
-  OutlierDetectorImpl(Cluster& cluster, Event::Dispatcher& dispatcher);
+  void onConsecutive5xx(HostPtr host);
+  Runtime::Loader& runtime() { return runtime_; }
 
   // Upstream::OutlierDetector
   void addChangedStateCb(ChangeStateCb cb) override { callbacks_.push_back(cb); }
 
+protected:
+  OutlierDetectorImpl(Cluster& cluster, Event::Dispatcher& dispatcher, Runtime::Loader& runtime,
+                      Stats::Store& stats, SystemTimeSource& time_source);
+
 private:
   void addHostSink(HostPtr host);
+  void armIntervalTimer();
+  void checkHostForUneject(HostPtr host, OutlierDetectorHostSinkImpl* sink, SystemTime now);
+  void ejectHost(HostPtr host);
+  static OutlierDetectionStats generateStats(const std::string& name, Stats::Store& store);
+  void onConsecutive5xxWorker(HostPtr host);
+  void onIntervalTimer();
+  void runCallbacks(HostPtr host);
 
+  Event::Dispatcher& dispatcher_;
+  Runtime::Loader& runtime_;
+  SystemTimeSource& time_source_;
+  OutlierDetectionStats stats_;
+  Event::TimerPtr interval_timer_;
```

Comment:
I named it like this because eventually it will be used for histogram computation, not just unejection.